staging/src/k8s.io/kube-aggregator/pkg/controllers/autoregister/autoregister_controller.go

```diff
@@ -0,0 +1,249 @@
+	for i := 0; i < threadiness; i++ {
+		// runWorker will loop until "something bad" happens.  The .Until will then rekick the worker
+		// after one second
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	// wait until we're told to stop
+	<-stopCh
+}
+
+func (c *autoRegisterController) runWorker() {
+	// hot loop until we're told to stop.  processNextWorkItem will automatically wait until there's work
+	// available, so we don't worry about secondary waits
+	for c.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem deals with one key off the queue.  It returns false when it's time to quit.
+func (c *autoRegisterController) processNextWorkItem() bool {
+	// pull the next work item from queue.  It should be a key we use to lookup something in a cache
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	// you always have to indicate to the queue that you've completed a piece of work
+	defer c.queue.Done(key)
+
+	// do your work on the key.  This method will contains your "do stuff" logic
+	err := c.syncHandler(key.(string))
+	if err == nil {
+		// if you had no error, tell the queue to stop tracking history for your key.  This will
+		// reset things like failure counts for per-item rate limiting
+		c.queue.Forget(key)
+		return true
+	}
+
+	// there was a failure so be sure to report it.  This method allows for pluggable error handling
+	// which can be used for things like cluster-monitoring
+	utilruntime.HandleError(fmt.Errorf("%v failed with : %v", key, err))
+	// since we failed, we should requeue the item to work on later.  This method will add a backoff
+	// to avoid hotlooping on particular items (they're probably still not going to work right away)
+	// and overall controller protection (everything I've done is broken, this controller needs to
+	// calm down or it can starve other useful work) cases.
+	c.queue.AddRateLimited(key)
+
+	return true
+}
+
+func (c *autoRegisterController) checkAPIService(name string) error {
+	desired := c.GetAPIServiceToSync(name)
+	curr, err := c.apiServiceLister.Get(name)
+
+	switch {
+	// we had a real error, just return it
+	case err != nil && !apierrors.IsNotFound(err):
+		return err
+
+	// we don't have an entry and we don't want one
+	case apierrors.IsNotFound(err) && desired == nil:
+		return nil
+
+	// we don't have an entry and we do want one
+	case apierrors.IsNotFound(err) && desired != nil:
+		_, err := c.apiServiceClient.APIServices().Create(desired)
+		return err
+
+	// we aren't trying to manage this APIService.  If the user removes the label, he's taken over management himself
+	case curr.Labels[AutoRegisterManagedLabel] != "true":
+		return nil
+
+	// we have a spurious APIService that we're managing, delete it
+	case desired == nil:
+		return c.apiServiceClient.APIServices().Delete(curr.Name, nil)
+
+		// if the spec's already match, nothing for us to do
+	case reflect.DeepEqual(curr.Spec, desired.Spec):
```

Comment:
> is reflect ok or do we need semantic?

Semantic isn't cleanly divorced from a global scheme yet. Reflect works for now.